upload/system/library/smailyforopencart/request.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace SmailyForOpenCart;
+
+class Request {
+    const HTTP_ERR_INVALID_SUBDOMAIN = 404;
+    const HTTP_ERR_UNAUTHORIZED = 401;
+
+    const API_ERR_INVALID_DATA = 203;
+    const API_ERR_FORBIDDEN = 202;
+    const API_ERR_WRONG_METHOD = 201;
+
+    protected $url = null;
+
+    protected $data = array();
+
+    protected $subdomain = null;
+
+    private $_username = null;
+
+    private $_password = null;
+
+    public function setSubdomain($subdomain) {
+        $this->subdomain = trim($subdomain);
+        $this->url = 'https://' . $this->subdomain . '.sendsmaily.net/api/';
+        return $this;
+    }
+
+    public function setCredentials($username, $password) {
+        $this->_username = trim($username);
+        $this->_password = trim($password);
+        return $this;
+    }
+
+    protected function compileUrl($endpoint) {
+        return $this->url . trim($endpoint, '/') . '.php';
+    }
+
+    public function get($endpoint, $data = array()) {
+        $url = $this->compileUrl($endpoint);
+
+        $ch = curl_init();
+        curl_setopt($ch, CURLOPT_URL, $url . '?' . http_build_query($data));
+        curl_setopt($ch, CURLOPT_HEADER, 0);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        curl_setopt($ch, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
+        curl_setopt($ch, CURLOPT_USERPWD, "{$this->_username}:{$this->_password}");
+
+        $api_call = curl_exec($ch);
+        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+
+        if ($http_code !== 200) {
+            throw new HTTPError('GET request to Smaily API failed', $http_code);
+        }
+
+        curl_close($ch);
```

Comment:
Please move `curl_close` before throwing an exception:

```suggestion
        $api_call = curl_exec($ch);
        $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($http_code !== 200) {
            throw new HTTPError('GET request to Smaily API failed', $http_code);
        }
```

Idea being that cURL resource will not be properly closed when exception is thrown. It might not really matter in this case, but a good thing to practice for closing resources connected to remote services (or servers; like databases).

Reply:
Thought to put `curl_close` near the end of code, but didn't consider that throwing an exception might end code execution earlier. Fixed in 8a9b501